File: src/ipc/messages/leapdatamodel.ts
import type { ValueType } from './valuetype';

export abstract class Value { }

export class StringValue extends Value {
    value: string;

    constructor(value: string) {
        super();
        this.value = value;
    }
}

export class IntegerValue extends Value {
    value: number;

    constructor(value: number) {
        super();
        this.value = value;
    }
}

export class FloatValue extends Value {
    value: number;

    constructor(value: number) {
        super();
        this.value = value;
    }
}

export class BooleanValue extends Value {
    value: boolean;

    constructor(value: boolean) {
        super();
        this.value = value;
    }
}

export class ListValue extends Value {
    value: Value[];
    valueType: ValueType;

    constructor(value: Value[], valueType: ValueType) {
        super();
        this.value = value;
        this.valueType = valueType;
    }
}

export class StructValue extends Value {
    value: Map<string, Value>;
    valueType: ValueType;

    constructor(value: Map<string, Value>, valueType: ValueType) {
        super();
        this.value = value;
        this.valueType = valueType;
    }
}

export class EnumValue extends Value {
    variant: string;
    value: StructValue;
    valueType: ValueType;

    constructor(variant: string, value: StructValue, valueType: ValueType) {
        super();
        this.variant = variant;
        this.value = value;
        this.valueType = valueType;
    }
}

File: src-tauri/src/ipc/messages/leap_data_model.rs
use super::value_type::ValueType;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub enum Value {
    String(StringValue),
    Integer(IntegerValue),
    Float(FloatValue),
    Boolean(BooleanValue),
    List(ListValue),
    Struct(StructValue),
    Enum(EnumValue),
}

impl Value {
    pub fn new_string(value: String) -> Self {
        Value::String(StringValue(value))
    }

    pub fn new_integer(value: i64) -> Self {
        Value::Integer(IntegerValue(value))
    }

    pub fn new_float(value: f64) -> Self {
        Value::Float(FloatValue(value))
    }

    pub fn new_boolean(value: bool) -> Self {
        Value::Boolean(BooleanValue(value))
    }

    pub fn new_list(value: Vec<Value>, value_type: ValueType) -> Self {
        Value::List(ListValue { value, value_type })
    }

    pub fn new_struct(value: HashMap<String, Value>, value_type: ValueType) -> Self {
        Value::Struct(StructValue { value, value_type })
    }

    pub fn new_enum(variant: String, value: StructValue, value_type: ValueType) -> Self {
        Value::Enum(EnumValue {
            variant,
            value,
            value_type,
        })
    }
}

#[derive(Debug, Clone)]
pub struct StringValue(pub String);

#[derive(Debug, Clone)]
pub struct IntegerValue(pub i64);

#[derive(Debug, Clone)]
pub struct FloatValue(pub f64);

#[derive(Debug, Clone)]
pub struct BooleanValue(pub bool);

#[derive(Debug, Clone)]
pub struct ListValue {
    pub value: Vec<Value>,
    pub value_type: ValueType,
}

#[derive(Debug, Clone)]
pub struct StructValue {
    pub value: HashMap<String, Value>,
    pub value_type: ValueType,
}

#[derive(Debug, Clone)]
pub struct EnumValue {
    pub variant: String,
    pub value: StructValue,
    pub value_type: ValueType,
}
